pkg/security/podsecuritypolicy/seccomp/strategy_test.go

```diff
@@ -161,49 +170,169 @@ func TestValidatePod(t *testing.T) {
 			pspAnnotations: allowSpecific,
 			podAnnotations: map[string]string{
-				api.SeccompPodAnnotationKey: "foo",
+				api.SeccompPodAnnotationKey: foo,
+			},
+			seccompProfile: &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &foo,
 			},
 			expectedError: "",
 		},
-		"invalid pod annotations, required profiles": {
+		"invalid pod annotations and no field, required profiles": {
 			pspAnnotations: allowSpecific,
 			podAnnotations: map[string]string{
-				api.SeccompPodAnnotationKey: "bar",
+				api.SeccompPodAnnotationKey: bar,
+			},
+			seccompProfile: nil,
+			expectedError:  "Forbidden: bar is not an allowed seccomp profile. Valid values are foo",
+		},
+		"invalid pod field and no annotations (localhost), required profiles": {
+			pspAnnotations: allowSpecific,
+			podAnnotations: nil,
+			seccompProfile: &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &bar,
 			},
 			expectedError: "Forbidden: bar is not an allowed seccomp profile. Valid values are foo",
 		},
-		"pod annotations, no required profiles": {
+		"invalid pod field and no annotations (runtime/default), required profiles": {
+			pspAnnotations: allowSpecific,
+			podAnnotations: nil,
+			seccompProfile: &api.SeccompProfile{
+				Type: api.SeccompProfileTypeRuntimeDefault,
+			},
+			expectedError: "Forbidden: RuntimeDefault is not an allowed seccomp profile. Valid values are foo",
+		},
+		"invalid pod field and no annotations (unconfined), required profiles": {
+			pspAnnotations: allowSpecific,
+			podAnnotations: nil,
+			seccompProfile: &api.SeccompProfile{
+				Type: api.SeccompProfileTypeUnconfined,
+			},
+			expectedError: "Forbidden: Unconfined is not an allowed seccomp profile. Valid values are foo",
+		},
+		"invalid pod field and annotations, required profiles": {
+			pspAnnotations: allowSpecific,
+			podAnnotations: map[string]string{
+				api.SeccompPodAnnotationKey: bar,
+			},
+			seccompProfile: &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &bar,
+			},
+			expectedError: "Forbidden: bar is not an allowed seccomp profile. Valid values are foo, pod.spec.securityContext.seccompProfile.type.localhostProfile: Forbidden: bar is not an allowed seccomp profile. Valid values are foo",
+		},
+		"pod annotations and no field, no required profiles": {
+			pspAnnotations: withoutSeccomp,
+			podAnnotations: map[string]string{
+				api.SeccompPodAnnotationKey: foo,
+			},
+			seccompProfile: nil,
+			expectedError:  "Forbidden: seccomp may not be set",
+		},
+		"pod field and no annotations, no required profiles": {
+			pspAnnotations: withoutSeccomp,
+			podAnnotations: nil,
+			seccompProfile: &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &foo,
+			},
+			expectedError: "Forbidden: seccomp may not be set",
+		},
```

Comment:
Does this mean that if a PSP exists and it does not specify seccomp annotations that seccomp may not be used on pods at all?

Reply:
Hm, this test cases were there before, so I think the validation should not be triggered at all if seccomp annotations are not available. Not exactly sure about this one…

Reply:
Very good point @hasheddan. The current documentation says that: 

> seccomp.security.alpha.kubernetes.io/allowedProfileNames - Annotation that specifies which values are allowed for the pod seccomp annotations. Specified as a comma-delimited list of allowed values. Possible values are those listed above, plus * to allow all profiles. **Absence of this annotation means that the default cannot be changed**.

So my take is that when `allowedProfileNames` is set to `nil` all workloads must use the profile set in `defaultProfileName`. If `defaultProfileName` is set to `unconfined`, this combination disables seccomp altogether. It may be worth changing a bit the test description to make that clearer.

Reply:
Hm alright I just added a note that this combination disables the usage of seccomp.

Reply:
@pjbgf @saschagrunert it makes sense to me that absence of `allowedProfileNames` means that you cannot change the default, but it also looks like the absence of both `allowedProfileNames` and `defaultProfileName` means that you cannot use seccomp profiles at all, which feels like it might not be the behavior we want considering that we can't require that these annotations be provided on a PSP like we can for some fields (i.e. `runAsUser`, `fsGroup`, etc.)

Reply:
So basically seccomp cannot be enabled if the PSP has:

`allowedProfileNames == nil && defaultProfileName == nil`
`allowedProfileNames == nil && defaultProfileName == 'unconfined'`
`allowedProfileNames == 'unconfined' && defaultProfileName == nil`
`allowedProfileNames == 'unconfined' && defaultProfileName == 'unconfined'`

I personally think we are better off documenting this and leaving as it is. 
Then group all changes we need from PSP into a new KEP. What do you think?

Reply:
> it also looks like the absence of both `allowedProfileNames` and `defaultProfileName` means that you cannot use seccomp profiles at all, which feels like it might not be the behavior we want

we should not change existing PSP behavior in this PR

Reply:
@liggitt would updating the PSP to have seccomp fields in spec be possible at all in the future, or is this API considered too mature at this point?

Reply:
> @liggitt would updating the PSP to have seccomp fields in spec be possible at all in the future, or is this API considered too mature at this point?

That's not clear at this point. It's less a question of maturity and more a question about the future of the PSP API and whether it is worth the effort/complexity of migrating an annotation to a field.

Reply:
Understood, thanks for the clarification!

Reply:
Sounds good, let's resolve this discussion for sanity reasons. :)